graspologic/embed/case.py

```diff
@@ -0,0 +1,198 @@
+            Contains the tuple (A, Y), where A is an adjacency matrix and Y is the
+            matrix of covariates.
+
+            A : array-like or networkx.Graph
+                Input graph to embed. See graspologic.utils.import_graph
+
+            Y : array-like, shape (n_vertices, n_covariates)
+                Covariate matrix. Each node of the graph is associated with a set of
+                `d` covariates. Row `i` of the covariates matrix corresponds to node
+                `i`, and the number of columns are the number of covariates.
+
+        y: Ignored
+
+        Returns
+        -------
+        self : object
+            Returns an instance of self.
+        """
+
+        # setup
+        if not isinstance(network, (tuple, list)):
+            msg = "Network should be a tuple-like object of (graph, covariates)."
+            raise TypeError(msg)
+        if len(network) != 2:
+            msg = "Network should be a tuple-like object of (graph, covariates)."
+            raise ValueError(msg)
+
+        graph, covariates = network
+        A = import_graph(graph)
+
+        # Create regularized Laplacian, scale covariates to unit norm
+        L = to_laplacian(A, form="R-DAD")
+        Y = covariates
+        if Y.ndim == 1:
+            Y = Y[:, np.newaxis]
+        Y = normalize(Y, axis=0)
+
+        # Use ratio of the two leading eigenvalues
+        # if alpha is None
+        self._get_tuning_parameter(L, Y)
+
+        # get embedding matrix as a LinearOperator (for computational efficiency)
+        n = A.shape[0]
+        mv, rmv = self._matvec(L, Y, a=self.alpha_, assortative=self.assortative)
+        L_ = LinearOperator((n, n), matvec=mv, rmatvec=rmv)
+
+        # Dimensionality reduction with SVD
+        self._reduce_dim(L_)
+
+        self.is_fitted_ = True
+        return self
+
+    def _get_tuning_parameter(self, L, Y):
+        """
+        Find the alpha which causes the leading eigenspace of LL and YYt to be the same.
+
+        Parameters
+        ----------
+        L : array
+            The regularized graph Laplacian.
+        Y : array
+            The covariate matrix.
+
+        Returns
+        -------
+        alpha : float
+            Tuning parameter which normalizes the leading eigenspace of LL and YYt.
+        """
+        # setup
+        if self.alpha is not None:
+            self.alpha_ = self.alpha
+            return self
+
+        # Laplacian leading eigenvector
+        (L_top,) = eigsh(L, return_eigenvectors=False, k=1)
+        if self.assortative:
+            L_top **= 2
+
+        # YY^T leading eigenvector
+        n = Y.shape[0]
+        YO = LinearOperator((n, n), matvec=lambda v: Y @ (Y.T @ v))
+        (YYt_top,) = eigsh(YO, return_eigenvectors=False, k=1)
+
+        # just use the ratio of the leading eigenvalues for the
+        # tuning parameter, or the closest value in its possible range.
+        self.alpha_ = np.float(L_top / YYt_top)
```

Comment:
why is this self.alpha_ not self.alpha?

Reply:
https://scikit-learn.org/stable/developers/develop.html#estimated-attributes

Reply:
I see, thanks